File: opendbc/can/packer.cc
#include <algorithm>
#include <cassert>
#include <cmath>
#include <stdexcept>
#include <utility>

#include "opendbc/can/common.h"


void set_value(std::vector<uint8_t> &msg, const Signal &sig, int64_t ival) {
  int i = sig.lsb / 8;
  int bits = sig.size;
  if (sig.size < 64) {
    ival &= ((1ULL << sig.size) - 1);
  }

  while (i >= 0 && i < msg.size() && bits > 0) {
    int shift = (int)(sig.lsb / 8) == i ? sig.lsb % 8 : 0;
    int size = std::min(bits, 8 - shift);

    msg[i] &= ~(((1ULL << size) - 1) << shift);
    msg[i] |= (ival & ((1ULL << size) - 1)) << shift;

    bits -= size;
    ival >>= size;
    i = sig.is_little_endian ? i+1 : i-1;
  }
}

CANPacker::CANPacker(const std::string& dbc_name) {
  dbc = dbc_lookup(dbc_name);
  assert(dbc);

  for (const auto& msg : dbc->msgs) {
    MessageData &msg_data = msg_lookup[msg.address];
    msg_data.msg_size = msg.size;

    for (const auto& sig : msg.sigs) {
      msg_data.signals[sig.name] = &sig;

      if (sig.name == "COUNTER") {
        msg_data.counter_signal = &sig;
      } else if (sig.name == "CHECKSUM") {
        msg_data.checksum_signal = &sig;
      }
    }
  }
}

std::vector<uint8_t> CANPacker::pack(uint32_t address, const std::vector<SignalPackValue> &signals) {
  auto msg_it = msg_lookup.find(address);
  if (msg_it == msg_lookup.end()) {
    LOGE("undefined address %d", address);
    return {};
  }

  MessageData &msg = msg_it->second;
  std::vector<uint8_t> ret(msg.msg_size, 0);

  // set all values for all given signal/value pairs
  bool counter_set = false;
  for (const auto& sigval : signals) {
    auto sig_it = msg.signals.find(sigval.name);
    if (sig_it == msg.signals.end()) {
      // TODO: do something more here. invalid flag like CANParser?
      LOGE("undefined signal %s - %d\n", sigval.name.c_str(), address);
      continue;
    }

    const Signal *sig = sig_it->second;
    int64_t ival = (int64_t)(round((sigval.value - sig->offset) / sig->factor));
    if (ival < 0) {
      ival = (1ULL << sig->size) + ival;
    }
    set_value(ret, *sig, ival);

    if (sig == msg.counter_signal) {
      msg.counter_value = sigval.value;
      counter_set = true;
    }
  }

  // set message counter if not already set
  if (msg.counter_signal && !counter_set) {
    set_value(ret, *msg.counter_signal, msg.counter_value);
    msg.counter_value = (msg.counter_value + 1) % (1 << msg.counter_signal->size);
  }

  // set message checksum
  if (msg.checksum_signal && msg.checksum_signal->calc_checksum) {
    unsigned int checksum = msg.checksum_signal->calc_checksum(address, *msg.checksum_signal, ret);
    set_value(ret, *msg.checksum_signal, checksum);
  }

  return ret;
}

// This function has a definition in common.h and is used in PlotJuggler
const Msg* CANPacker::lookup_message(uint32_t address) {
  return dbc->addr_to_msg.at(address);
}

File: opendbc/can/packer.py
import math

from mypy_extensions import mypyc_attr

from opendbc.car.carlog import carlog
from opendbc.can._types import Signal, SignalType


def set_value(msg: bytearray, sig: Signal, ival: int) -> None:
  sig_lsb: int = sig.lsb
  sig_size: int = sig.size
  sig_le: bool = sig.is_little_endian
  i: int = sig_lsb // 8
  bits: int = sig_size
  if sig_size < 64:
    ival &= (1 << sig_size) - 1
  msg_len: int = len(msg)
  while 0 <= i < msg_len and bits > 0:
    shift: int = sig_lsb % 8 if (sig_lsb // 8) == i else 0
    size: int = min(bits, 8 - shift)
    mask: int = ((1 << size) - 1) << shift
    msg[i] &= ~mask
    msg[i] |= (ival & ((1 << size) - 1)) << shift
    bits -= size
    ival >>= size
    i = i + 1 if sig_le else i - 1


@mypyc_attr(allow_interpreted_subclasses=True)
class CANPacker:
  def __init__(self, dbc_name: str) -> None:
    from opendbc.can.dbc import DBC
    self.dbc = DBC(dbc_name)
    self.counters: dict[int, int] = {}

  def pack(self, address: int, values: dict[str, float]) -> bytearray:
    msg = self.dbc.addr_to_msg.get(address)
    if msg is None:
      carlog.error(f"msg not found for {address=}")
      return bytearray()
    dat: bytearray = bytearray(msg.size)
    counter_set: bool = False
    for name, value in values.items():
      sig = msg.sigs.get(name)
      if sig is None:
        carlog.error(f"unknown signal {name=} in {msg.name}")
        continue
      ival: int = int(math.floor((value - sig.offset) / sig.factor + 0.5))
      if ival < 0:
        ival = (1 << sig.size) + ival
      set_value(dat, sig, ival)
      if sig.type == SignalType.COUNTER or sig.name == "COUNTER":
        self.counters[address] = int(value)
        counter_set = True
    sig_counter: Signal | None = next((s for s in msg.sigs.values() if s.type == SignalType.COUNTER or s.name == "COUNTER"), None)
    if sig_counter is not None and not counter_set:
      if address not in self.counters:
        self.counters[address] = 0
      set_value(dat, sig_counter, self.counters[address])
      self.counters[address] = (self.counters[address] + 1) % (1 << sig_counter.size)
    sig_checksum: Signal | None = next((s for s in msg.sigs.values() if s.type > SignalType.COUNTER), None)
    if sig_checksum is not None and sig_checksum.calc_checksum is not None:
      checksum: int = sig_checksum.calc_checksum(address, sig_checksum, dat)
      set_value(dat, sig_checksum, checksum)
    return dat

  def make_can_msg(self, name_or_addr: str | int, bus: int, values: dict[str, float]) -> tuple[int, bytes, int]:
    if isinstance(name_or_addr, int):
      addr: int = name_or_addr
    else:
      msg = self.dbc.name_to_msg.get(name_or_addr)
      if msg is None:
        carlog.error(f"msg not found for {name_or_addr=}")
        return 0, b'', bus
      addr = msg.address
    dat: bytearray = self.pack(addr, values)
    if len(dat) == 0:
      return 0, b'', bus
    return addr, bytes(dat), bus
